### PR TITLE
feat: partial checkpoint delete [DET-9491]

### PR DESCRIPTION
**Pull request by NicholasBlaskey**
## Description

Allows partial checkpoint deletion through sdk or added cli command.

## Test Plan

unit + integration + e2e tests

manual testing

Create experiment
```
det e create examples/tutorials/mnist_pytorch/const.yaml examples/tutorials/mnist_pytorch/ -f
```

Get checkpoint id
```
det e lc $exp_id --csv
```

Partial delete no op
```
det checkpoint rm 663013e4-1234-476b-a151-6b7aa77ece87
```

Should see (wait for the task to finish by it falling off ```det task``` and use ```det e lc $exp_id --csv```)
```Trial ID,# of Batches,State,Validation Metric,UUID,Resources,Size
26068,937,COMPLETED,0.0497250406331161,663013e4-1234-476b-a151-6b7aa77ece87,"code/
code/README.md
code/adaptive.yaml
code/const.yaml
code/data.py
code/dist_random.yaml
code/distributed.yaml
code/layers.py
code/metrics_spam.py~
code/model_def.py
code/simple_script.py~
load_data.json
metadata.json
state_dict.pth
trial_state.pkl",13.8MB
```

Remove `load_data.json` manually from file system and start partial delete no op
```
rm /tmp/determined-cp/663013e4-1234-476b-a151-6b7aa77ece87/load_data.json
det checkpoint rm 663013e4-1234-476b-a151-6b7aa77ece87
```

Should see
```
Trial ID,# of Batches,State,Validation Metric,UUID,Resources,Size
26068,937,PARTIALLY_DELETED,0.0497250406331161,663013e4-1234-476b-a151-6b7aa77ece87,"code/
code/README.md
code/adaptive.yaml
code/const.yaml
code/data.py
code/dist_random.yaml
code/distributed.yaml
code/layers.py
code/metrics_spam.py~
code/model_def.py
code/simple_script.py~
metadata.json
state_dict.pth
trial_state.pkl",13.8MB
```

Remove code directory
```
det checkpoint rm 663013e4-1234-476b-a151-6b7aa77ece87 --glob "code/*"
```

Should see
```
Trial ID,# of Batches,State,Validation Metric,UUID,Resources,Size
26068,937,PARTIALLY_DELETED,0.0497250406331161,663013e4-1234-476b-a151-6b7aa77ece87,"code/
metadata.json
state_dict.pth
trial_state.pkl",13.8MB

```

Fully delete checkpoint
```
det checkpoint rm 663013e4-1234-476b-a151-6b7aa77ece87 --glob "trial_state.pkl" --glob "state_dict.pth" --glob "code" --glob "metadata.json"
```

Should see
```
Trial ID,# of Batches,State,Validation Metric,UUID,Resources,Size
26068,937,DELETED,0.0497250406331161,663013e4-1234-476b-a151-6b7aa77ece87,"code/
metadata.json
state_dict.pth
trial_state.pkl",13.8MB

```
## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/common/storage/hdfs.py [hunk 1]
```diff
@@ -61,6 +61,7 @@ def download(
         self.client.download(src, dst, overwrite=True)
 
     @util.preserve_random_state
-    def delete(self, tgt: str) -> None:
+    def delete(self, tgt: str, globs: List[str]) -> Dict[str, int]:
```
Comment (NicholasBlaskey) [inline harness/determined/common/storage/hdfs.py:64]: got the go ahead to remove hdfs next release, will do it in a PR which will be up soon

File: harness/determined/common/storage/shared.py [hunk 2]
```diff
@@ -169,18 +170,51 @@ def restore_path(
...
+
+        to_delete_dirs = {}
+        to_delete_files = {}
+        for file_glob in globs:
+            for path in glob.glob(f"{storage_dir}/{file_glob}", recursive=True):
+                if os.path.isfile(path):
+                    to_delete_files[path] = True
+                elif os.path.isdir(path):
+                    to_delete_dirs[path] = True
+                # TODO should we delete system links too?
```
Comment (NicholasBlaskey) [inline harness/determined/common/storage/shared.py:198]: is system links a potential here?

File: harness/determined/common/storage/shared.py [hunk 3]
```diff
@@ -169,18 +170,51 @@ def restore_path(
...
+        for path in to_delete_dirs:
+            util.rmtree_nfs_safe(path, ignore_errors=False)
+
+        resources = self._list_directory(storage_dir)
+
+        # TODO we upload this "metadata.json" file in checkpoints sometimes AFTER
+        # we calculate resources so resources won't contain it.
+        # This actually might be an issue if we sometimes have it in resources?
+        # todo this might be an issue.
+        if "metadata.json" in resources:
```
Comment (NicholasBlaskey) [inline harness/determined/common/storage/shared.py:212]: this and other `metadata.json` hack is needed since we often create this file after we calculate and store checkpoint resources in the database. 

This causes us when just checking if the checkpoint matches the database state to think that it always doesn't.

Not sure if this handling of it is fine.

File: harness/determined/common/storage/shared.py [hunk 4]
```diff
@@ -169,18 +170,51 @@ def restore_path(
...
+            for path in glob.glob(f"{storage_dir}/{file_glob}", recursive=True):
+                if os.path.isfile(path):
+                    to_delete_files[path] = True
+                elif os.path.isdir(path):
+                    to_delete_dirs[path] = True
+                # TODO should we delete system links too?
+
+        # Delete files first then delete paths.
+        for path in to_delete_files:
+            os.remove(path)  # TODO do we need an os.remove safe here?
```
Comment (NicholasBlaskey) [inline harness/determined/common/storage/shared.py:202]: do we need an `os.remove` safe that retries like how `rmtree_nfs_safe` does?

File: master/internal/api_checkpoint.go [hunk 5]
```diff
@@ -197,24 +167,169 @@ func (a *apiServer) DeleteCheckpoints(
...
+		return nil, status.Errorf(codes.InvalidArgument, "converting checkpoint: %s", cErr)
+	}
+
+	for _, g := range req.CheckpointGlobs {
+		if len(g) == 0 {
+			// Avoid weirdness where someone passes in "" then we concat {uuid}/{glob}
+			// and we unexpectedly delete the whole checkpoint folder.
+			return nil, status.Errorf(codes.InvalidArgument, "cannot have empty string glob")
+		}
+		if strings.Contains(g, "..") {
```
Comment (NicholasBlaskey) [inline master/internal/api_checkpoint.go:302]: is this sufficient to make sure no one passes a glob that could delete all of shared_fs?

File: master/pkg/tasks/task_gc.go [hunk 6]
```diff
@@ -61,12 +65,15 @@ func (g GCCkptSpec) ToTaskSpec() TaskSpec {
 					0o600,
 					tar.TypeReg,
 				),
+				// TODO: Do we use this file? I don't think so???
+				// If we do we need to update add glob
 				g.Base.AgentUserGroup.OwnedArchiveItem(
 					"checkpoint_gc/checkpoints_to_delete.json",
-					[]byte(jsonify(g.ToDelete)),
```
Comment (NicholasBlaskey) [inline master/pkg/tasks/task_gc.go:66]: do we still use these files? As far as I can tell we pass everything through command line arguments now?

**Review by ioga — COMMENTED** [review]
I read the python half; will come back for the rest later.

File: harness/determined/cli/checkpoint.py [hunk 7]
```diff
@@ -123,7 +123,28 @@ def delete_checkpoints(args: Namespace) -> None:
...
+            "in checkpoint storage. Do you still want to proceed?"
+        )
+    ):
+        c_uuids = args.checkpoints_uuids.split(",")
+        remove_body = bindings.v1CheckpointsRemoveFilesRequest(
+            checkpointGlobs=args.glob,
+            checkpointUuids=c_uuids,
+        )
+        bindings.post_CheckpointsRemoveFiles(cli.setup_session(args), body=remove_body)
+        print(f"Removal of files from checkpoints {args.checkpoints_uuids} is in progress")
```
[the remark this answers is not in the document]
Comment (NicholasBlaskey) [reply harness/determined/cli/checkpoint.py:145]: I believe the only way to wait for checkpoint gc now is to poll `/tasks` like we do in checkpoint tests now, which feels somewhat hacky trying to catch the timing. Since delete checkpoint doesn't have a wait option would a follow up ticket be okay? 

I think we do need to overall improve checkpoint gc visibility to the user somehow.

File: e2e_tests/tests/cluster/test_checkpoints.py [hunk 8]
```diff
@@ -391,6 +391,157 @@ def test_delete_experiment_with_no_checkpoints() -> None:
...
+                break
+    else:
+        pytest.fail("did not find two checkpoints in state completed")
+
+    s = bindings.get_GetExperiment(
+        test_session,
+        experimentId=exp_id,
+    ).experiment.checkpointSize
+    assert s is not None
+    starting_size = int(s)
```
[the remark this answers is not in the document]
Comment (NicholasBlaskey) [reply e2e_tests/tests/cluster/test_checkpoints.py:432]: Yeah it is unfortunate, not sure if an easy fix in binding generation code

created https://hpe-aiatscale.atlassian.net/browse/DET-9511

File: master/internal/api_checkpoint.go (hunk 5 above)
Comment (ioga) [reply master/internal/api_checkpoint.go:302]: I can't think of other checks we could do here in go...

however in python, we could check `os.path.common_path(checkpoint_dir, normpath(globbed_path)) == checkpoint_dir` for extra safety.

another wild idea (and a lot of work) is to spawn a python worker per checkpoint uuid, and have each worker `os.chroot` into their respective uuid directory...

File: master/static/migrations/20230523115842_add-checkpoint-partial-delete-state.tx.up.sql [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  exec_text := format('CREATE VIEW checkpoints_old_view AS %s', checkpoints_old_view);
+  execute exec_text;
+  exec_text := format('CREATE VIEW checkpoints_new_view AS %s', checkpoints_new_view);
+  execute exec_text;
+  exec_text := format('CREATE VIEW checkpoints AS %s', checkpoints);
+  execute exec_text;
+  exec_text := format('CREATE VIEW checkpoints_view AS %s', checkpoints_view);
+  execute exec_text;
+  exec_text := format('CREATE VIEW proto_checkpoints_view AS %s', proto_checkpoints_view);
+  execute exec_text;
```
Comment (ioga) [inline master/static/migrations/20230523115842_add-checkpoint-partial-delete-state.tx.up.sql:47]: nice!

File: master/pkg/tasks/task_gc.go (hunk 6 above)
Comment (rb-determined-ai) [reply master/pkg/tasks/task_gc.go:66]: you do need this file, because otherwise there are system limit on how many or how long the cli args can be.

If we're relying on cli args at the moment, then it's a regression.  I originally added this file to get around that limitation.

Comment (NicholasBlaskey) [reply master/pkg/tasks/task_gc.go:66]: Checkpoint storage is still transferred as a file, I'm not sure if it was determined that checkpoint storage is the argument at risk of hitting the system limit or if it was a regression.

This PR awhile back which added delete checkpoints switched passing checkpoint uuids as a list instead of a file and never deleted the file. 
https://github.com/determined-ai/determined/pull/4246/files#diff-ebcdb45e3a87d3548c878e57fab45c41041815fec850f14fe74a755b1b2bc7bdR74

I'll just switch back to json file for uuids.

File: master/internal/api_checkpoint.go (hunk 5 above)
Comment (NicholasBlaskey) [reply master/internal/api_checkpoint.go:302]: Good idea on common path check, added it.

**Comment by NicholasBlaskey** [comment]
@ioga implemented the `metadata.json` changes a little out of my depth on the `_checkpoint.py` but hopefully looks good

**Comment by stoksc** [comment]
@ioga did you review go stuff too?

**Comment by ioga** [comment]
@stoksc yes
